sdks/go/pkg/beam/core/runtime/harness/statemgr_test.go

```diff
@@ -0,0 +1,248 @@
+	recvErr error
+	// Blocks the write routine
+	send    chan *pb.StateRequest
+	sendErr error
+}
+
+func (f *fakeStateClient) Recv() (*pb.StateResponse, error) {
+	// Blocks until something is sent.
+	return <-f.recv, f.recvErr
+}
+
+func (f *fakeStateClient) Send(req *pb.StateRequest) error {
+	f.send <- req // blocks until consumed.
+	return f.sendErr
+}
+
+func TestStateChannel(t *testing.T) {
+	// The logging of channels closed is quite noisy for this test
+	log.SetOutput(ioutil.Discard)
+
+	expectedError := fmt.Errorf("EXPECTED ERROR")
+
+	tests := []struct {
+		name              string
+		caseFn            func(t *testing.T, c *StateChannel, client *fakeStateClient) error
+		expectedErr       error
+		validateCancelled bool
+	}{
+		{
+			name: "HappyPath",
+			caseFn: func(t *testing.T, c *StateChannel, client *fakeStateClient) error {
+				// Verify that we can send a bunch of requests and get the expected paired response back.
+				// This is not a real StateAPI protocol, since that's handled at a higher level than the StateChannel,
+				// but validates that the routing occurs correctly.
+				const count = 10
+				for i := 0; i < count; i++ {
+					go func() {
+						req := <-client.send
+						client.recv <- &pb.StateResponse{
+							Id: req.Id, // Ids need to match up to ensure routing can occur properly.
+							Response: &pb.StateResponse_Get{
+								Get: &pb.StateGetResponse{
+									ContinuationToken: req.GetGet().GetContinuationToken(),
+								},
+							},
+						}
+					}()
+				}
+				for i := 0; i < count; i++ {
+					token := []byte(fmt.Sprintf("%d", i))
+					resp, err := c.Send(&pb.StateRequest{
+						Request: &pb.StateRequest_Get{
+							Get: &pb.StateGetRequest{
+								ContinuationToken: token,
+							},
+						},
+					})
+					if err != nil {
+						t.Fatalf("unexpected error from Send: %v", err)
+					}
+					if got, want := string(resp.GetGet().GetContinuationToken()), string(token); got != want {
+						t.Fatalf("req/response mismatch from Send: got %v, want %v", got, want)
+					}
+				}
+				return nil
+			},
+		}, {
+			name: "readEOF",
+			caseFn: func(t *testing.T, c *StateChannel, client *fakeStateClient) error {
+				go func() {
+					req := <-client.send // Send should succeed.
+
+					client.recvErr = io.EOF
+					client.recv <- &pb.StateResponse{
+						Id: req.Id,
+					}
+				}()
+				_, err := c.Send(&pb.StateRequest{})
+				return err
+			},
+			expectedErr:       io.EOF,
+			validateCancelled: true,
+		}, {
+			name: "readOtherErr",
+			caseFn: func(t *testing.T, c *StateChannel, client *fakeStateClient) error {
+				go func() {
+					req := <-client.send // Send should succeed.
+
+					client.recvErr = expectedError
+					client.recv <- &pb.StateResponse{
+						Id: req.Id,
+					}
+				}()
+				_, err := c.Send(&pb.StateRequest{})
+				return err
+			},
+			expectedErr:       expectedError,
+			validateCancelled: true,
+		}, {
+			name: "readResponseChannelDeleted",
```

Comment:
This test case gave me some trouble to understand, compared to the others. Is this one meant to check the case in StateChannel.read where there is no consumer for the state response? (Line 332 in statemgr.go)

Reply:
Correct.